### PR TITLE
Themes: Decouple rendering from element creation

### PR DESCRIPTION
**Pull request by ockham**
This is one more step toward isomorphic routing and supersedes #3207.
Client-side rendering is moved from `themes` to a generic module, and is called via `client/sections` (as generated by `server/bundler/loader`, and invoked by `client/boot`). This way, we'll be able to call theme specific middleware on the server to only create React elements, which we will then `renderToString()`.

This PR also moves some stuff to more generic locations (`renderElements` to `client/controller`, `client-side-effects` to `components/`) for better reusability.

To test -- verify that everything still works, especially `/design` (single-site, multi-site, logged-out, and selecting any theme's `Details`, and back).

props @mcsf for fixing a nasty middleware related bug causing lots of flashes

### CODE REVIEW

**Comment by mcsf** [comment]
@ockham: feel free to check off the last bug from the description if the latest commits helped.

**Comment by mcsf** [comment]
Carrying over @ockham's comment on 3b68e72 just in case we rebase that commit out of existence:

> >  plus bonus for eliminating any LoC from the `loader.js` metacode.
> 
> Totally agree with that. Not so sure about making `page` calls less explicit here, since they're already all over this file. What about the alternative
> 
> ```js
> require( ' + JSON.stringify( module ) + ' )( page, controller.renderElements );'
> ```
> 
> with that second argument function invoked inside `my-sites/themes` after elements have been created (much like it's now on `master`), as discussed in Slack? Arguments to that would then be `router` and `renderer`, which I find kind of appealing, since it captures quite nicely client and server specific differences. (We can totally rename `renderElements` to `clientRenderer`, BTW).

While I'd generally be in favor of decoupling things where they make sense (i.e. router and renderer are conceptually and semantically different), I wonder if the extra cost of exposing a larger set of separate concepts to other Calypso developers might shadow the inherent benefits of that decoupling. What motivates you most to decouple them? Do we envision scenarios where we don't have 1:1 relationships between our {client,server} routers and our {client,server} renderers?

So it's clear: I'm not necessarily defending the current implementation, just asking the questions.

**Comment by ockham** [comment]
> I wonder if the extra cost of exposing a larger set of separate concepts to other Calypso developers might shadow the inherent benefits of that decoupling. What motivates you most to decouple them?

Mmhm, good point. I think mostly that for me it seems like `clientRouter` is kinda shadowing what's really just another `page` call (but I have to go to another file to look that up, meh), while the renderer doesn't have any precedent in this file. For me personally I think it's easier to keep routing and rendering conceptually apart -- the `page` call inside `clientRenderer` along with the mw destructuring kind of distracted me from the `renderElements` invocation...

> Do we envision scenarios where we don't have 1:1 relationships between our {client,server} routers and our {client,server} renderers?

Probably not :-)

**Comment by ockham** [comment]
Scratch passing `clientRenderer` to the chunk, btw -- the chunk wouldn't know which route to pass it to.

It's still unfortunate we can't allow catch-all route definitions like I had in here, cause the same constraint will affect all other sections, too -- if people want to go isomorphic, not only will they need to decouple their element generation and rendering, they will also have to give up catch-all routes for e.g. `navigation` and `siteSelection` (which I personally find quite handy)

The only way out would be, again, a `page` call in the loader right below the chunk invocation, but this time, it'd have to make sure to be only called if the chunk is not rendering itself. That could be deduced from a `routing: 'isomorphic'` flag in `sections.js`, tho. Benefit: less structure imposed on individual chunks.

**Comment by ockham** [comment]
Oh well, since I can't even cook up a working PoC, I'm punting my suggestion.

File: client/controller.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	page( route, ...[ ...mws, renderElements ] );
+}
+
+export function renderElements( context ) {
+	renderPrimary( context );
+	renderSecondary( context );
+}
+
+function renderPrimary( context ) {
+	const { path, primary } = context;
```
Comment (ockham) [inline client/controller.js]: This fixes https://github.com/Automattic/wp-calypso/pull/3354/files#r53517871

File: client/controller.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * Internal dependencies
+ */
+import page from 'page';
+import debugFactory from 'debug';
+
+const debug = debugFactory( 'calypso:controller' );
+
+export function clientRouter( route, ...mws ) {
```
Comment (ehg) [inline client/controller.js]: s/mws/middlewares please :)

File: client/my-sites/themes/index.js [hunk 3]
```diff
@@ -1,42 +1,34 @@
...
 		Object.keys( routes ).forEach( route => {
-			page( route, ...routes[ route ] );
+			router( route, ...routes[ route ] );
 		} )
-		page( '/design*', renderPrimary ); // Call explicitly here because client-specific
 	}
 	if ( config.isEnabled( 'manage/themes/details' ) ) {
-		page( '/themes/:slug/:site_id?', details );
-		page( '/themes*', renderPrimary ); // Call explicitly here because client-specific
+		router( '/themes/:slug/:site_id?', details );
```
Comment (ehg) [inline client/my-sites/themes/index.js]: I guess this should be in the `routes` object?

File: client/controller.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+export function renderElements( context ) {
+	renderPrimary( context );
+	renderSecondary( context );
+}
+
+function renderPrimary( context ) {
+	const { path, primary } = context;
+
+	// FIXME: temporary hack until we have a proper isomorphic, one tree
+	// routing solution. Do NOT do this!
```
Comment (ehg) [inline client/controller.js]: Maybe link #3302 here, since that's where the good stuff will happen.

**Comment by ehg** [comment]
Saw this:
screen shot 2016-02-22 at 13 27 51

To reproduce:
1. View http://calypso.localhost:3000/design logged-out
2. Click a `Popover` details link
3. Refresh the themes sheet page
4. Press browser back to go to `/design`
5. Press browser forward to go to the themes sheet

File: client/controller.js (hunk 4 above)
Comment (ockham) [inline client/controller.js]: Like so? e4adc4b26df97726ec5252aa32f9d8d741d42d7c

Comment (ehg) [inline client/controller.js]: :+1:

File: client/controller.js (hunk 2 above)
Comment (ockham) [inline client/controller.js]: > s/mws/middlewares please :)

b4eaccb01c9ccf22f045f47384d6cc273a9e59d7

File: client/my-sites/themes/index.js (hunk 3 above)
Comment (ockham) [inline client/my-sites/themes/index.js]: a688cf5affd695e5bfed9c4064db70a42b886780

**Comment by ockham** [comment]
Ugh. Item 3, refreshing the sheet, is likely crucial here, as it SSR's `LayoutLoggedOutDesign` with some sheet-specific args. Will investigate.

**Comment by ockham** [comment]
> To reproduce:
1. View http://calypso.localhost:3000/design logged-out
2. Click a Popover details link
3. Refresh the themes sheet page
4. Press browser back to go to /design
5. Press browser forward to go to the themes sheet

@ehg BTW this currently also produces an error with `master`, albeit a different one :-(

`Uncaught Error: Invariant Violation: findComponentRoot(..., .aggvzg3mdc.2.0.0.1.0.1.0.1.$/=10.0): Unable to find element. This probably means the DOM was unexpectedly mutated (e.g., by the browser), usually due to forgetting a <tbody> when using tables, nesting tags like <form>, <p>, or <a>, or using non-SVG elements in an <svg> parent. Try inspecting the child nodes of the element with React ID `.aggvzg3mdc.2.0`.`

**Comment by ehg** [comment]
> BTW this currently also produces an error with master, albeit a different one :-(

Most likely due to this: https://github.com/Automattic/wp-calypso/pull/3455 — proper layout last routing can't come soon enough :)

**Comment by ockham** [comment]
> Most likely due to this: #3455 — proper layout last routing can't come soon enough :)

Tried with the check added back already, no luck :-(

**Comment by ockham** [comment]
Noting here that

1. Since this branch has the same error as `master`, it's not a regression
2. This is only enabled in `dev` and `wpcalypso`, but neither in `staging` nor `production`
3. Theme search is tested to work

**Comment by ehg** [comment]
:+1: after squash. Code looks good, and correct behaviour observed in SS, MS, LO — other than aforementioned placeholdery bug.

**Pull request merged**